### PR TITLE
fix(filter): handle UTF-8 chunk splitting, CRLF, and data: prefix in SSE parser

### PR DESCRIPTION
**Pull request by leseb**
## Summary

The `anthropic_stream_events` SSE parser had three correctness bugs:

- **UTF-8 chunk splitting**: HTTP chunks can split multibyte UTF-8 characters mid-sequence — the filter now buffers incomplete trailing bytes and reassembles them with the next chunk
- **CRLF line endings**: the parser only recognized `\n\n` boundaries, so valid SSE using `\r\n` or standalone `\r` was buffered forever — line endings are now normalized with a deferred-CR strategy to avoid false boundaries when `\r\n` splits across chunks
- **`data:` prefix**: the parser required `data: ` (with space), but the SSE spec says the space is optional — `data:value` is now accepted

Also adds a configurable `max_partial_event_bytes` limit for the line buffer, schema-complete Anthropic output fields (`stop_details`, `container`, extended usage), and end-of-stream flush for pending buffers.

## Test plan

- [x] `split_utf8_character_buffered_across_chunks` — verifies multi-byte char split across chunks reassembles correctly
- [x] `crlf_event_boundaries_parsed` — verifies `\r\n\r\n` delimiters are recognized
- [x] `crlf_split_across_chunks_not_false_boundary` — verifies trailing `\r` doesn't create a premature event
- [x] `data_without_space_after_colon_accepted` — verifies `data:value` parses
- [x] `done_sentinel_without_space_after_colon` — verifies `data:[DONE]` completes stream
- [x] 9 additional edge-case tests (invalid UTF-8, truncated end-of-stream, oversized buffer, deferred CR flush, etc.)
- [x] All 14 existing + new tests pass, `make lint` clean

### CODE REVIEW

**Comment by leseb** [comment]
Manual validation full flow, you need an inference backend, i used vllm locally on 8000:

```
cargo run -p praxis-proxy -- -c examples/configs/ai/anthropic/messages-to-openai.yaml
```

Then execute a request:

```
curl -N http://127.0.0.1:8080/v1/messages \
  -H "Content-Type: application/json" \
  -H "x-api-key: test" \
  -d '{"model":"Qwen/Qwen3-0.6B","max_tokens":100,"messages":[{"role":"user","content":"Hi"}],"stream":true}'
event: message_start
data: {"message":{"container":null,"content":[],"id":"msg_18bdd0d183ffe7a8","model":"Qwen/Qwen3-0.6B","role":"assistant","stop_details":null,"stop_reason":null,"stop_sequence":null,"type":"message","usage":{"cache_creation":null,"cache_creation_input_tokens":null,"cache_read_input_tokens":null,"inference_geo":null,"input_tokens":0,"output_tokens":0,"server_tool_use":null,"service_tier":null}},"type":"message_start"}

event: content_block_start
data: {"content_block":{"text":"","type":"text"},"index":0,"type":"content_block_start"}

event: content_block_delta
data: {"delta":{"text":"","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"<think>","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"\n","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"Okay","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":",","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" the","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" user","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" said","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" \"","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"Hi","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"\".","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" I","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" need","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" to","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" respond","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" appropriately","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":".","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" Since","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" they","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" started","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" the","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" conversation","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" with","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" a","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" greeting","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":",","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" I","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" should","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" acknowledge","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" it","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":".","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" Maybe","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" say","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" \"","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"Hello","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"!\"","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" to","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" keep","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" it","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" friendly","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":".","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" I","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" should","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" check","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" if","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" they","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" need","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" help","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":",","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" but","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" keep","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" it","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" open","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"-ended","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":".","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" Let","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" me","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" make","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" sure","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" the","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" response","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" is","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" simple","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" and","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" welcoming","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":".","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" Alright","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":",","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" that","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" should","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" do","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" it","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":".\n","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"</think>","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"\n\n","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"Hello","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"!","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" How","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" can","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" I","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" assist","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" you","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" today","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"?","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":" 😊","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_delta
data: {"delta":{"text":"","type":"text_delta"},"index":0,"type":"content_block_delta"}

event: content_block_stop
data: {"index":0,"type":"content_block_stop"}

event: message_delta
data: {"delta":{"container":null,"stop_details":null,"stop_reason":"end_turn","stop_sequence":null},"type":"message_delta","usage":{"cache_creation_input_tokens":null,"cache_read_input_tokens":null,"input_tokens":null,"output_tokens":0,"server_tool_use":null}}

event: message_stop
data: {"type":"message_stop"}
```

**Review by franciscojavierarceo — APPROVED** [review]

File: filter/src/builtins/http/ai/anthropic/stream_events/mod.rs [hunk 1]
```diff
@@ -152,28 +171,175 @@ impl HttpFilter for AnthropicStreamEventsFilter {
...
+    max_partial_event_bytes: usize,
+) -> Result<Option<Bytes>, FilterError> {
+    let combined = combine_pending_utf8(ctx, bytes);
+    let Some(valid_up_to) = valid_utf8_prefix_len(ctx, combined.as_slice(), end_of_stream) else {
+        return Ok(Some(passthrough_with_line_buffer(ctx, combined)));
+    };
+    let Some(valid_bytes) = combined.as_slice().get(..valid_up_to) else {
+        return Ok(None);
+    };
+    let Some(chunk_str) = std::str::from_utf8(valid_bytes).ok() else {
```
[the remark this answers is not in the document]
Comment (leseb) [reply filter/src/builtins/http/ai/anthropic/stream_events/mod.rs:199]: over-engineered solution? also the repo forbids usage of `unsafe`?

File: docs/filters/http/ai/anthropic_stream_events.md [hunk 2]
```diff
@@ -7,8 +7,23 @@ Transforms streaming SSE responses between `OpenAI` and Anthropic formats, proce
 
 Requires Cargo feature: `ai-inference`.
 
-## Example
+## Configuration
+
+| Field | Type | Required | Description |
```
[the remark this answers is not in the document]
Comment (leseb) [reply docs/filters/http/ai/anthropic_stream_events.md:12]: out of scope, will fix separately

File: filter/src/builtins/http/ai/anthropic/stream_events/config.rs [hunk 3]
```diff
@@ -14,8 +19,39 @@ use serde::Deserialize;
...
+    Ok(cfg)
+}
+
+/// Validate the maximum partial SSE event byte limit.
+fn validate_max_partial_event_bytes(value: usize) -> Result<(), FilterError> {
+    if value == 0 {
+        return Err("anthropic_stream_events: 'max_partial_event_bytes' must be greater than 0".into());
+    }
+
+    if value > MAX_JSON_BODY_BYTES {
```
[the remark this answers is not in the document]
Comment (leseb) [reply filter/src/builtins/http/ai/anthropic/stream_events/config.rs:49]: 04e4b21b

File: filter/src/builtins/http/ai/anthropic/stream_events/mod.rs [hunk 4]
```diff
@@ -694,17 +994,378 @@ mod tests {
         );
     }
 
+    #[test]
+    fn split_utf8_character_buffered_across_chunks() {
+        let (filter, mut ctx) = make_filter_and_context();
+
+        // € (Euro sign) encodes as [0xE2, 0x82, 0xAC] in UTF-8.
```
[the remark this answers is not in the document]
Comment (leseb) [reply filter/src/builtins/http/ai/anthropic/stream_events/mod.rs]: 04e4b21b

File: docs/filters/http/ai/anthropic_stream_events.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (leseb) [reply docs/filters/http/ai/anthropic_stream_events.md:12]: https://github.com/praxis-proxy/praxis/pull/751
